File: apple/Sources/ReactNativeWatchOS/Modifiers/GlassEffect.swift
import SwiftUI

/// SwiftUI `.glassEffect(_:in:)` — watchOS 26 "Liquid Glass". Applies a
/// Liquid Glass material to the view, clipped to a shape.
///
/// `variant` selects the `Glass` material (`regular` / `clear` / `identity`),
/// `shape` the clip (`capsule` / `rect` / `circle` / `roundedRectangle`,
/// the last reading `cornerRadius`). Defaults: `.regular` in `.capsule`.
///
/// Gated to watchOS 26+ — returns the view unchanged on older OS, since
/// these APIs don't exist before "Liquid Glass".
enum RNWGlassEffectModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("glassEffect") { view, params, _ in
            guard #available(watchOS 26.0, *) else { return view }

            let glass = makeGlass(params.string("variant"))
            let shape = params.string("shape") ?? "capsule"
            let cornerRadius = params.cgFloat("cornerRadius") ?? 0

            // `glassEffect(_:in:)` takes `some Shape`, so the shape can't be
            // type-erased without losing its identity for the morph system.
            // Branch per concrete shape instead.
            switch shape {
            case "rect":
                return AnyView(view.glassEffect(glass, in: Rectangle()))
            case "circle":
                return AnyView(view.glassEffect(glass, in: Circle()))
            case "roundedRectangle":
                return AnyView(view.glassEffect(
                    glass,
                    in: RoundedRectangle(cornerRadius: cornerRadius)
                ))
            default:
                return AnyView(view.glassEffect(glass, in: Capsule()))
            }
        }
    }

    @available(watchOS 26.0, *)
    private static func makeGlass(_ variant: String?) -> Glass {
        switch variant {
        case "clear":    return .clear
        case "identity": return .identity
        default:         return .regular
        }
    }
}

File: src/modifiers/glassEffect.ts
import { createModifier } from './createModifier';

/// The `Glass` material variant. `regular` is the default frosted glass;
/// `clear` is a more transparent variant; `identity` is a no-op pass-through
/// (useful for conditionally disabling the effect while keeping layout stable).
export type GlassVariant = 'regular' | 'clear' | 'identity';

/// The clipping shape the glass is rendered into. `roundedRectangle` reads
/// the optional `cornerRadius` param (defaults to 0).
export type GlassShape = 'capsule' | 'rect' | 'circle' | 'roundedRectangle';

export interface GlassEffectParams {
  /// Glass material variant. Defaults to `regular`.
  variant?: GlassVariant;
  /// Shape the glass is clipped to. Defaults to `capsule`.
  shape?: GlassShape;
  /// Corner radius — only used when `shape` is `roundedRectangle`.
  cornerRadius?: number;
}

/// SwiftUI `.glassEffect(_:in:)` (watchOS 26 "Liquid Glass"). Applies a
/// Liquid Glass background to the view, clipped to `shape`. No-op on
/// watchOS < 26 (the native side returns the view unchanged).
export function glassEffect(params: GlassEffectParams = {}) {
  return createModifier('glassEffect', params);
}
